optimade/adapters/structures/pymatgen.py

```diff
@@ -126,3 +131,85 @@ def _pymatgen_species(
         pymatgen_species.append(dict(zip(chemical_symbols, concentration)))
 
     return pymatgen_species
+
+
+def from_pymatgen(pmg_structure: Structure) -> StructureResourceAttributes:
+    """Convert a pymatgen `Structure` (3D) into an OPTIMADE `StructureResourceAttributes` model.
+
+    Parameters:
+        pmg_structure: The pymatgen `Structure` to convert.
+
+    Returns:
+        An OPTIMADE `StructureResourceAttributes` model, which can be converted to a raw Python
+            dictionary with `.dict()` or to JSON with `.json()`.
+
+    """
+
+    if not isinstance(pmg_structure, Structure):
+        raise RuntimeError(
+            f"Cannot convert type {type(pmg_structure)} into an OPTIMADE `StructureResourceAttributes` model."
+        )
+
+    attributes = {}
+    attributes["cartesian_site_positions"] = pmg_structure.lattice.get_cartesian_coords(
+        pmg_structure.frac_coords
+    ).tolist()
+    attributes["lattice_vectors"] = pmg_structure.lattice.matrix.tolist()
+    attributes["species_at_sites"] = [_.symbol for _ in pmg_structure.species]
+    attributes["species"] = [
+        {"name": _.symbol, "chemical_symbols": [_.symbol], "concentration": [1]}
+        for _ in set(pmg_structure.composition.elements)
+    ]
+    attributes["dimension_types"] = [1, 1, 1]
+    attributes["nperiodic_dimensions"] = 3
+    attributes["nelements"] = len(pmg_structure.composition.elements)
+    attributes["chemical_formula_anonymous"] = _pymatgen_anonymized_formula_to_optimade(
+        pmg_structure.composition
+    )
+    attributes["elements"] = sorted(
+        [_.symbol for _ in pmg_structure.composition.elements]
+    )
+    attributes["chemical_formula_reduced"] = _pymatgen_reduced_formula_to_optimade(
+        pmg_structure.composition
+    )
+    attributes["chemical_formula_descriptive"] = pmg_structure.composition.formula
+    attributes["elements_ratios"] = [
+        pmg_structure.composition.get_atomic_fraction(e) for e in attributes["elements"]
+    ]
+    attributes["nsites"] = len(attributes["species_at_sites"])
+
+    attributes["last_modified"] = None
+    attributes["immutable_id"] = None
+    attributes["chemical_formula_descriptive"] = None
```

Comment:
Why are you overwriting the "chemical_formula_descriptive" here when you just set the value on line 175 ?

Reply:
I couldn't decide on this one (and I can't remember which line I added last 😅). I'm not sure if the pymatgen formula is that useful as `chemical_formula_descriptive`, but it probably doesn't make a huge difference to use it.

```suggestion
```